File: fxgl-base/src/main/java/com/almasb/fxgl/gameplay/AchievementEvent.java
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.gameplay;

import javafx.event.Event;
import javafx.event.EventType;

/**
 * Occurs on achievement unlocked.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
public class AchievementEvent extends Event {

    public static final EventType<AchievementEvent> ANY =
            new EventType<>(Event.ANY, "ACHIEVEMENT_EVENT");

    public static final EventType<AchievementEvent> ACHIEVED =
            new EventType<>(ANY, "ACHIEVED");

    private Achievement achievement;

    public AchievementEvent(Achievement achievement) {
        this(ANY, achievement);
    }

    public AchievementEvent(EventType<? extends AchievementEvent> eventType, Achievement achievement) {
        super(eventType);
        this.achievement = achievement;
    }

    /**
     * @return achievement associated with the event
     */
    public Achievement getAchievement() {
        return achievement;
    }

    @Override
    public String toString() {
        return "AchievementEvent[name=" + achievement.getName()
                + ",description= " + achievement.getDescription() + "]";
    }
}

File: fxgl-base/src/main/kotlin/com/almasb/fxgl/gameplay/achievement/AchievementEvent.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.gameplay.achievement

import javafx.event.Event
import javafx.event.EventType

/**
 * Occurs on achievement unlocked.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
open class AchievementEvent(
        eventType: EventType<out AchievementEvent>,
        val achievement: Achievement
) : Event(eventType) {

    companion object {
        @JvmField val ANY = EventType<AchievementEvent>(Event.ANY, "ACHIEVEMENT_EVENT")

        @JvmField val ACHIEVED = EventType(ANY, "ACHIEVED")
    }

    constructor(achievement: Achievement) : this(ANY, achievement) {}

    override fun toString() =
            "AchievementEvent[name=${achievement.name}, description=${achievement.description}]"
}